traceconfig: generate an annotated Markdown document with source refs

pillo79 opened the pull request:
This pull request adds a new `traceconfig` build target to generate Markdown documentation for Kconfig symbols that details the reason a value was chosen for each symbol in the configuration. The list of symbols is split in three sorted groups:

* **Visible symbols**: parameters that have a value and are user-configurable
* **Invisible symbols**: parameters that have a value which are hidden by default (often target of `select`s or implicit `default`s).
* **Unset symbols**: parameters that do not have a value assigned

Symbols that have a value include a readable and clickable (if the viewer supports it) link to the location of the statement that resulted in that value being applied, be it an explicit assignment or a `default`. For the symbols that are `select`-ed or `imply`-ed by multiple sources, the set of matching requirements are printed.

[Documentation](https://builds.zephyrproject.io/zephyr/pr/95808/docs/build/kconfig/tracing.html) for the new feature is available and a [test suite](https://github.com/zephyrproject-rtos/zephyr/actions/runs/17795626467/job/50582879272#step:13:414) has been added to validate Kconfig source reference functionality.

## Sample generated documentation

The new target generates a `zephyr/kconfig-trace.md` file in the project's build folder. [This](https://github.com/user-attachments/files/22383650/kconfig-trace.md) is an example of a full generated documentation for the `sample/drivers/spi_flash` project running on the `arduino_opta//m7` board. 

The objective is for the generated document to be easily readable by humans with a simple editor. The entries are rendered like this:

### As a raw source

```
| Type | Name                             |      Value |  Source  | Location                                                                                                                                  |
|-----:|:---------------------------------|-----------:|:--------:|:------------------------------------------------------------------------------------------------------------------------------------------|
| bool | `CONFIG_SERIAL`                  |          y | user set | [boards/arduino/opta/arduino\_opta\_stm32h747xx\_m7\_defconfig:27](<../../boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig#L27>) |
| bool | `CONFIG_SIZE_OPTIMIZATIONS`      |          y | default  | [Kconfig.zephyr:485](<../../Kconfig.zephyr#L485>)                                                                                         |
| bool | `CONFIG_SOC_EARLY_INIT_HOOK`     |          y | selected | `SOC_SERIES_STM32H7X && SOC_FAMILY_STM32`                                                                                                 |
| bool | `CONFIG_SOC_FLASH_STM32`         |          y | default  | [drivers/flash/Kconfig.stm32:22](<../../drivers/flash/Kconfig.stm32#L22>)                                                                 |
|  hex | `CONFIG_SRAM_BASE_ADDRESS`       | 0x24000000 | default  | [arch/Kconfig:228](<../../arch/Kconfig#L228>)                                                                                             |
| bool | `CONFIG_SRAM_REGION_PERMISSIONS` |          y | selected | `ARM_MPU && CPU_HAS_MPU`                                                                                                                  |
|  int | `CONFIG_SRAM_SIZE`               |        512 | default  | [arch/Kconfig:220](<../../arch/Kconfig#L220>)                                                                                             |
| bool | `CONFIG_STACK_ALIGN_DOUBLE_WORD` |          y | default  | [arch/arm/core/Kconfig:152](<../../arch/arm/core/Kconfig#L152>)                                                                           |
| bool | `CONFIG_STDOUT_CONSOLE`          |          y | user set | [samples/drivers/spi\_flash/prj.conf:1](<../../samples/drivers/spi_flash/prj.conf#L1>)                                                    |
```

### As a Github table

Note: links here have been edited to refer to the upstream used in the PR generation. They are usually relative to where the file was generated.

| Type | Name | Value | Source | Location |
|-------:|:----------------------------------------------------|--------------------------:|:--------:|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| bool | `CONFIG_SERIAL`                  |          y | user set | [boards/arduino/opta/arduino\_opta\_stm32h747xx\_m7\_defconfig:27](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/boards/arduino/opta/arduino_opta_stm32h747xx_m7_defconfig#L27>) |
| bool | `CONFIG_SIZE_OPTIMIZATIONS`      |          y | default  | [Kconfig.zephyr:485](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/Kconfig.zephyr#L485>)                                                                                         |
| bool | `CONFIG_SOC_EARLY_INIT_HOOK`     |          y | selected | `SOC_SERIES_STM32H7X && SOC_FAMILY_STM32`                                                                                                 |
| bool | `CONFIG_SOC_FLASH_STM32`         |          y | default  | [drivers/flash/Kconfig.stm32:22](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/drivers/flash/Kconfig.stm32#L22>)                                                                 |
|  hex | `CONFIG_SRAM_BASE_ADDRESS`       | 0x24000000 | default  | [arch/Kconfig:228](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/arch/Kconfig#L228>)                                                                                             |
| bool | `CONFIG_SRAM_REGION_PERMISSIONS` |          y | selected | `ARM_MPU && CPU_HAS_MPU`                                                                                                                  |
|  int | `CONFIG_SRAM_SIZE`               |        512 | default  | [arch/Kconfig:220](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/arch/Kconfig#L220>)                                                                                             |
| bool | `CONFIG_STACK_ALIGN_DOUBLE_WORD` |          y | default  | [arch/arm/core/Kconfig:152](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/arch/arm/core/Kconfig#L152>)                                                                           |
| bool | `CONFIG_STDOUT_CONSOLE`          |          y | user set | [samples/drivers/spi\_flash/prj.conf:1](<../../zephyr/blob/7d84a3b639526374eee405a06c773b572b74c913/samples/drivers/spi_flash/prj.conf#L1>)                                                    |

### Rendered in VSCode:

<img width="1257" height="384" alt="image" src="https://github.com/user-attachments/assets/42219de5-5a4e-4389-96f9-9c3a5cc8d95d" />

## Review notes

* Kconfiglib already used `loc`s (a tuple of filename and line number) in a few specific places to store line information. 
* A refactor was first applied to expand the usage to all statements modifying the value, with the same `loc` element  shared with multiple entities. 
* The *source reference* concept is then added to `Symbols` to collapse this information in a common form, which is exported for later usage by external tools. 
* A processor that reads this information and writes the user-visible Markdown document is then added, along with a test suite for the functionality.

pillo79 commented on tests/kconfig/srcrefs/validate_srcrefs.py:
```diff
@@ -0,0 +1,150 @@
+            return False
+    else:
+        # source references, trim input path
+        if not isinstance(actual[SYM_LOC], list | tuple) or len(actual[SYM_LOC]) != 2:
```
I had originally `(list, tuple)` which I found easier, but `ruff` complained and requested the pipe syntax :-(

pdgendt replied:
We have 3.10 as minimal python version, using the `|` operator is fine.

mathieuchopstm commented on scripts/kconfig/traceconfig.py:
```diff
@@ -0,0 +1,119 @@
+
+    args = parser.parse_args()
+
+    with open(args.dotconfig_file + '-srcrefs.pickle', 'rb') as f:
```
Also, my understanding is that *we are the ones generating this pickle* so it shouldn't be an issue.

mathieuchopstm reviewed (commented):
Seems fairly reasonable.

kartben commented on scripts/kconfig/traceconfig.py:
```diff
@@ -0,0 +1,119 @@
+import pickle
+
+import kconfiglib
+from tabulate import tabulate
```
yes

tejlmand commented on scripts/kconfig/kconfiglib.py:
changes to Kconfiglib should go here first https://github.com/zephyrproject-rtos/Kconfiglib
That repo is now considered the official upstream Kconfiglib.

If changes to kconfig lib are needed fast, like in this PR, then this commit can reference the upstream PR.

pillo79 replied:
Thanks, didn't know that, will clean up this PR a bit and post there first.

mathieuchopstm commented on scripts/kconfig/traceconfig.py:
```diff
@@ -0,0 +1,119 @@
+
+
+def source_link(refpath, fn, ln):
+    if refpath:
+        fn = os.path.relpath(fn, refpath)
+
+    link_fn = fn
+    disp_fn = os.path.normpath(fn).replace("../", "").replace("_", "\\_")
```
Hmm, so this *removes* the `../`?! Doesn't that mean that the resulting path is not correct? What I was thinking about is that IIRC, `pathlib` has a method to "canonicalize" a path (e.g., `/zephyrproject/zephyr/../modules` --> `/zephyrproject/modules`)

pillo79 replied:
>  so this removes the `../`?! Doesn't that mean that the resulting path is not correct?

Yes and no 🙂 the input paths to the function are saved in absolute form by `kconfiglib`, they get converted here to an already minimal relative path (with `../` only at the beginning), and then I strip away all of them _for display purposes_ (note that `link_fn` is unchanged). Links look clean and are really intuitive, you know where those paths are!

mathieuchopstm replied:
Ok, makes sense. I guess sticking to `os.path` could be fine in this case then.

mathieuchopstm commented on scripts/kconfig/kconfiglib.py:
Ok. Thanks for the clarification :+1:

tejlmand reviewed (commented):
Looks good.

But please note that changes to Kconfiglib should go upstream first: https://github.com/zephyrproject-rtos/Kconfiglib and then brought down to Zephyr with a reference.

Note also that kconfiglib is used by other projects.

It even states that Kconfiglib is python2/3 compliant (perhaps we should remove python2 as it's no longer maintained, but guess that's another discussion :wink: )

erwango commented on doc/build/kconfig/tracing.rst:
```diff
@@ -0,0 +1,59 @@
+``traceconfig`` target can be used to generate a file in the build directory
+that details how each symbol got its final value.
+
+After building the Zephyr project as usual, the Kconfig trace can be generated
+with either of these commands:
+
+   .. code-block:: bash
+
+      west build -t traceconfig
+
+   .. code-block:: bash
+
+      ninja traceconfig
```
Any possibility to have this done automatically, for instance by setting a preference in `.west/config` ?

pillo79 committed:
kconfiglib: node: add 'loc' attribute

Replace 'filename' and 'linenr' usages with a single 'loc' attribute,
which is a (filename, linenr) tuple. This simplifies code dealing with
locations, and makes it explicitly constant. The original attributes are
now provided via properties for compatibility.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
kconfiglib: refactor Kconfig._warn() to use 'loc'

Replace filename and linenr parameters to _warn() with a single loc
element, which is a constant (filename, linenr) tuple that can be more
efficiently passed and stored.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
kconfiglib: add 'user_loc' to Symbol and Choice classes

The 'user_loc' stores the location (filename and line number) where a
symbol or choice was last set via a direct user selection.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
kconfiglib: add 'loc' to reverse dependencies

Store the location (filename and line number) where a 'select', 'imply',
'range' or 'default' was added to a Symbol or Choice.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
kconfiglib: track origin for symbol values

Track information about what caused a value to be set in the 'origin'
property of Symbol. 'imply' and 'select' dependencies do not have a
location associated with them, so in those cases the location is
a string representation of the dependency expression. For defaults and
user values, the location in the Kconfig file is stored.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
kconfig: collect and save trace data

Collect and save trace data for all symbols in the merged configuration.
This includes information about where each symbol was defined or
assigned, which can be useful for debugging complex Kconfig setups.

The trace data includes the following information for each symbol:
- Name
- Visibility
- Type
- Value
- Kind and location of value origin (as defined in kconfiglib)

The trace data is saved for later use in two formats: a binary pickle
file and a human-readable JSON file.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
traceconfig: add target to export value origin traces to Markdown

The new 'traceconfig' target generates a Markdown file listing all
configuration symbols, their values, and where those values originated
(user assignment, default, selection, implication, or unset).

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
tests/kconfig: add Kconfig origin tracing test

Add a test case that verifies Kconfig value origin traces are correctly
generated in both JSON and pickle formats. The test includes a sample
Kconfig configuration with dependencies, selections, and choices to
cover various scenarios. A validation script checks the generated source
origin data against expected values.

In addition, the 'traceconfig' target is called to test the generation
of the Markdown file with the Kconfig trace report.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
**v2**:
- Renamed the concept of "srcrefs" to "origin" (fingers and eyes are a lot happier!)  
- Reworked kconfiglib changes into zephyrproject-rtos/kconfiglib#18 (first 5 commits). Now, only the `origin` information is computed in the common kconfiglib, bundling of all other fields and the export code is in `scripts/kconfig.py` (in Zephyr) - this cut looks cleaner to me than it was before.
- No change to final output.
- All comments but auto-run addressed.

@erwango Tried to look into adding optional auto-run via `.west/config`, but I would have to import west stuff into "kconfig target" code. Not sure if it is OK to do so? If not, should I rework `traceconfig` to be a west command? Any advice on other ways to optionally run the doc generation at every instance? TIA

erwango commented:
> @erwango Tried to look into adding optional auto-run via .west/config, but I would have to import west stuff into "kconfig target" code. Not sure if it is OK to do so? If not, should I rework traceconfig to be a west command? Any advice on other ways to optionally run the doc generation at every instance? TIA

Tbh, I won't help much here. @carlescufi maybe ?

pillo79 commented:
@tejlmand and others, please re-review, would be useful to have in 4.3...
